clients/drcachesim/tests/cache_replacement_policy_unit_test.cpp

```diff
@@ -0,0 +1,133 @@
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL GOOGLE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+// Unit tests for cache replacement policies
+#include <iostream>
+#include <assert.h>
+#include "cache_replacement_policy_unit_test.h"
+#include "simulator/cache_lru.h"
+
+class cache_lru_test_t : public cache_lru_t {
+public:
+    void
+    initialize_cache(int associativity, int line_size, int total_size)
+    {
+        caching_device_stats_t *stats = new cache_stats_t(line_size, "", true);
+        if (!init(associativity, line_size, total_size, nullptr, stats, nullptr)) {
+            std::cerr << "LRU cache failed to initialize"
+                      << "\n";
+            exit(1);
+        }
+    }
+
+    void
+    unit_test_replace_which_way()
+    {
+        // Create and initialize a 4-way set associative cache with line size of 32 and
+        // total size of 256 bytes.
+        initialize_cache(4, 32, 256);
+
+        memref_t ref_1;
+        ref_1.data.type = TRACE_TYPE_READ;
+        ref_1.data.size = 1;
+        // Access the cache line in the following fashion. This sequence follows the
+        // sequence shown in https://github.com/DynamoRIO/dynamorio/issues/4881.
+        // Access the first row.
+        ref_1.data.addr = 0;
+        request(ref_1); // This accesses "a" in issue 4881.
+
+        ref_1.data.addr = 64;
+        request(ref_1); // This accesses "b" in issue 4881.
+
+        ref_1.data.addr = 128;
+        request(ref_1); // This accesses "c" in issue 4881.
+
+        ref_1.data.addr = 192;
+        request(ref_1); // This accesses "d" in issue 4881.
+
+        // After the above accesses, the counters for each way will be as follows:
+        //  way 0 ("a" in issue 4881): 3
+        //  way 1 ("b" in issue 4881): 2
+        //  way 2 ("c" in issue 4881): 1
+        //  way 3 ("d" in issue 4881): 0
+        // At this point way 0 ("a") has the highest counter so it should be replaced by
+        // the LRU policy.
+        assert(replace_which_way(0) == 0);
+
+        ref_1.data.addr = 0;
+        request(ref_1); // This replaces way 0 ("a").
+        // At this point way 0 has been replaced(accessed) and way 1 has the highest
```

Comment:
> "way 1 has the highest counter"

Not sure thinking in terms of the counters inside the implementation is the right way to approach this test: this goes back to the black box test suggestion.  Which line is the least recently used is what we care about, not how that is computed internally.  If we invert the counters internally so 0 is old and 3 is new, it will still work properly, and the test should still apply.

Reply:
Understood. Updated the comments and added helper methods.